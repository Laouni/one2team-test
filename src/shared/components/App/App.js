// @flow

import React from 'react';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import { stocksCac40NasdaqData } from '../../api/stocks';
import StocksDataTable from '../StocksDataTable/StocksDataTable';
import StocksGraphic from '../StocksGraphic/StocksGraphic';

import './App.css';

/**
 * App
 */
class App extends React.Component {
  state = {
    stocksList: [],
  };

  /**
   * componentDidMount
   */
  async componentDidMount(): void {
    await this.loadStocks();
    // TODO: comment this line to stop display continous data
    // setInterval(this.loadStocks, 1000);
  }

  /**
   * return latest stocks
   * @returns {Promise<void>}
   */
  loadStocks = async () => {
    try {
      const stocksList = await stocksCac40NasdaqData();

      this.setState({ stocksList });
    } catch (error) {
      console.log(error);
    }
  };

  /**
   * Get a simplified list of stocks
   * @returns {Array<Object>}
   */
  getSimplifiedStocksList = () => {
    const { stocksList } = this.state;
    let index = 1;
    return stocksList.map(obj => ({
      index: index++,
      NASDAQ: obj.stocks.NASDAQ.toFixed(2),
      CAC40: obj.stocks.CAC40.toFixed(2),
    }));
  };

  /**
   * render
   * @returns {HTMLElement}
   */
  render() {
    const data = this.getSimplifiedStocksList();

    return (
      <div className="App">
        <Grid container spacing={3}>
          <Grid item xs={12}>
            <Typography className="App-Typography" variant="h4" component="h2" gutterBottom>
              One2Team - Test Perfomed By Beauclair BILONG
            </Typography>
          </Grid>
          <Grid item xs={4}>
            <div className="App-DataTable">
              <StocksDataTable data={data} />
            </div>
          </Grid>
          <Grid item xs={8}>
            <div className="App-Graphic">
              <StocksGraphic data={data} />
            </div>
          </Grid>
        </Grid>
      </div>
    );
  }
}

export default App;
