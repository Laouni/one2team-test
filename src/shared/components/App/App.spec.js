import React from 'react';
import renderer from 'react-test-renderer';
import Axios from 'axios';
import { shallow } from '../../../../tools/config/test/enzyme';
import App from './App';

jest.mock('../StocksDataTable/StocksDataTable', () => 'mock-stocks-data-table');
jest.mock('../StocksGraphic/StocksGraphic', () => 'mock-stocks-graphic');

describe('App component', () => {
  it('should render correctly', () => {
    const component = renderer.create(<App />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call API if stocks list is not defined', async () => {
    Axios.get = jest.fn().mockReturnValue(
      Promise.resolve({
        data: [
          {
            timestamp: 1572179500328,
            index: 3854,
            stocks: {
              NASDAQ: 2.671635251444507e-53,
              CAC40: 1.4047233574433293e-62,
            },
          },
          {
            timestamp: 1572179501328,
            index: 3855,
            stocks: {
              NASDAQ: 2.7260407904981827e-53,
              CAC40: 1.3051173150328128e-62,
            },
          },
          {
            timestamp: 1572179502333,
            index: 3856,
            stocks: {
              NASDAQ: 2.7330589021584643e-53,
              CAC40: 1.3993285768808365e-62,
            },
          },
          {
            timestamp: 1572179503334,
            index: 3857,
            stocks: {
              NASDAQ: 2.896808355539025e-53,
              CAC40: 2.207972762403556e-62,
            },
          },
          {
            timestamp: 1572179504334,
            index: 3858,
            stocks: {
              NASDAQ: 2.8993316629164897e-53,
              CAC40: 2.2105438319963443e-62,
            },
          },
        ],
      }),
    );

    const component = shallow(<App />);

    await component.instance().loadStocks();

    expect(component.instance().state.stocksList.length).toEqual(5);
  });
});
