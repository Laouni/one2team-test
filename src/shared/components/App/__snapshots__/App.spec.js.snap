// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render correctly 1`] = `
<div
  className="App"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h2
        className="MuiTypography-root App-Typography MuiTypography-h4 MuiTypography-gutterBottom"
      >
        One2Team - Test Perfomed By Beauclair BILONG
      </h2>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
    >
      <div
        className="App-DataTable"
      >
        <mock-stocks-data-table
          data={Array []}
        />
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
    >
      <div
        className="App-Graphic"
      >
        <mock-stocks-graphic
          data={Array []}
        />
      </div>
    </div>
  </div>
</div>
`;
